python/graphstorm/model/lm_embed.py

```diff
@@ -28,109 +29,200 @@
-        # Note: self.lm_emb_cache is initialized by calling warmup
-        for ntype, idx in input_nodes.items():
-            if ntype in lm_emb_cache:
-                lm_feats[ntype] = lm_emb_cache[ntype][idx].to(dev)
-    else:
-        # TODO: Release the bert cache properly
-        #       This may need support from DistDGL
-        # Need bert training
-        for (lm_ntypes, lm_node_feats), lm_model \
-            in zip(lm_models_info, lm_models):
-            input_ntypes = []
-            input_lm_feats = {}
+    def __init__(self, g, node_lm_configs, num_train, lm_infer_batch_size):
+        super(LMModels, self).__init__()
+        # More than one node type may share the same BERT model.
+        # To avoid duplicate BERT models, we save add the BERT model
+        # in the module dict once, whose key is the node types of the BERT model.
+        # We then maintain a mapping from the node type to the key
+        # to help find the right BERT model for a node type.
+        self._lm_map = {}
+        self._lm_models = nn.ModuleDict()
+        self._lm_node_feats = {}
+        for lm_config in node_lm_configs:
+            lm_model = init_lm_model(lm_config,
+                                     num_train=num_train,
+                                     lm_infer_batch_size=lm_infer_batch_size)
+            # A list of node types sharing the same lm model
+            lm_ntypes = lm_config["node_types"]
+            lm_node_feats = get_lm_node_feats(g, lm_model, lm_ntypes)
+            for ntype, feats in lm_node_feats.items():
+                assert ntype not in self._lm_node_feats, \
+                        f"More than one BERT model runs on Node {ntype}."
+                self._lm_node_feats[ntype] = feats
+            key = str(lm_ntypes)
+            self._lm_models[key] = lm_model
             for ntype in lm_ntypes:
+                self._lm_map[ntype] = key
+
+    def forward(self, input_nodes, lm_emb_cache=None):
+        """ Do language model forward pass on input_nodes
+
+        Parameters
+        ----------
+        input_nodes: dict
+            Input nodes for different node types
+        lm_emb_cache: dict
+            Language model embedding cache for different node types
+        """
+        lm_feats = {}
+
+        # Get the device from lm_models
+        # The cached BERT embedding should be moved to the same device
+        # as lm_models.
+        dev = self.device
+        if lm_emb_cache is not None:
+            # No bert training, Get cached LM embedding
+            # Note: self.lm_emb_cache is initialized by calling warmup
+            for ntype, idx in input_nodes.items():
+                if ntype in lm_emb_cache:
+                    lm_feats[ntype] = lm_emb_cache[ntype][idx].to(dev)
+        else:
+            # TODO: Release the bert cache properly
+            #       This may need support from DistDGL
+            # Need bert training
+            for ntype in self.ntypes:
```

Comment:
If multiple ntypes share the same lm_model, can we batch them together to improve efficiency?

Reply:
does it matter? I tested it. there is no difference for batch size of 32, 128, 1024.
also, how can we batch bert computation on different node types? the length of text sequences on different node types can be different.